hudi-common/src/main/java/org/apache/hudi/common/schema/HoodieSchemaComparatorForSchemaEvolution.java

```diff
@@ -0,0 +1,346 @@
+  }
+
+  protected boolean schemaEqualsInternal(HoodieSchema s1, HoodieSchema s2) {
+    if (s1 == s2) {
+      return true;
+    }
+    if (s1 == null || s2 == null) {
+      return false;
+    }
+    if (s1.getType() != s2.getType()) {
+      return false;
+    }
+
+    switch (s1.getType()) {
+      case RECORD:
+        return recordSchemaEquals(s1, s2);
+      case ENUM:
+        return enumSchemaEquals(s1, s2);
+      case ARRAY:
+        return arraySchemaEquals(s1, s2);
+      case MAP:
+        return mapSchemaEquals(s1, s2);
+      case FIXED:
+        return fixedSchemaEquals(s1, s2);
+      case UNION:
+        return unionSchemaEquals(s1, s2);
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case NULL:
+      case DATE:
+        // DATE is INT with date logical type (no additional properties)
+      case UUID:
+        // UUID is STRING with uuid logical type (no additional properties)
+        return true;
+      case DECIMAL:
+        return decimalSchemaEquals(s1, s2);
+      case TIME:
+        return timeSchemaEquals(s1, s2);
+      case TIMESTAMP:
+        return timestampSchemaEquals(s1, s2);
+      default:
+        throw new IllegalArgumentException("Unknown schema type: " + s1.getType());
+    }
+  }
+
+  protected boolean validateRecord(HoodieSchema s1, HoodieSchema s2) {
+    return s1.isError() == s2.isError();
+  }
+
+  private boolean recordSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    if (!validateRecord(s1, s2)) {
+      return false;
+    }
+
+    List<HoodieSchemaField> fields1 = s1.getFields();
+    List<HoodieSchemaField> fields2 = s2.getFields();
+
+    if (fields1.size() != fields2.size()) {
+      return false;
+    }
+
+    for (int i = 0; i < fields1.size(); i++) {
+      if (!fieldEquals(fields1.get(i), fields2.get(i))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  protected boolean validateField(HoodieSchemaField f1, HoodieSchemaField f2) {
+    if (!f1.name().equals(f2.name())) {
+      return false;
+    }
+
+    if (f1.order() != f2.order()) {
+      return false;
+    }
+
+    // Check if both have default values
+    if (f1.hasDefaultValue() != f2.hasDefaultValue()) {
+      return false;
+    }
+
+    // If both have default values, they must be equal
+    return !f1.hasDefaultValue() || f1.defaultVal().get().equals(f2.defaultVal().get());
+  }
+
+  private boolean fieldEquals(HoodieSchemaField f1, HoodieSchemaField f2) {
+    if (!validateField(f1, f2)) {
+      return false;
+    }
+
+    return schemaEqualsInternal(f1.schema(), f2.schema());
+  }
+
+  protected boolean enumSchemaEquals(HoodieSchema s1, HoodieSchema s2) {
+    // Check name equality first
+    if (!s1.getName().equals(s2.getName())) {
+      return false;
+    }
+
+    List<String> symbols1 = s1.getEnumSymbols();
+    List<String> symbols2 = s2.getEnumSymbols();
+
+    // Quick size check before creating sets
```

Comment:
nit: Comment is outdated mentioning `sets`, but `List` is used.

Reply:
Fixed comment.